# Data analysis and visualization

This script contains the code for analyzing and visualizing the data from the UNHCR datasets.

***

### Packages and settings

In [1]:
import pandas as pd
import pycountry
import warnings

In [2]:
pd.options.display.float_format = '{:.3f}'.format # Avoid scientific notation
pd.options.mode.chained_assignment = None  # default='warn'
warnings.simplefilter(action='ignore', category=FutureWarning)

*********

## Data sources

### 1. UNHCR refugees and asylum seekers estimates

In [3]:
UNHCR = pd.read_csv("../data/unhcr-refugees/population.csv")

In [4]:
UNHCR.tail()

Year Country of origin Country of origin (ISO) Country of asylum  \
68  2019                 -                       -                 -   
69  2020                 -                       -                 -   
70  2021                 -                       -                 -   
71  2022                 -                       -                 -   
72  2023                 -                       -                 -   

   Country of asylum (ISO)  Refugees under UNHCR's mandate  Asylum-seekers  \
68                       -                        20414669         4148141   
69                       -                        20661846         4184926   
70                       -                        21327285         4616135   
71                       -                        29429078         5442319   
72                       -                        31637408         6858499   

    Returned refugees  IDPs of concern to UNHCR  Returned IDPss  \
68             317181                  43503362         5343793   
69             250951                  48557439         3184118   
70             429234                  51322623         5265622   
71            1356261                  57321197         8324166   
72            1052074                  63251367         5092064   

    Stateless persons  Others of concern  \
68            4217774            3857179   
69            4179331            3939756   
70            4338192            4223095   
71            4428314            6008804   
72            4358188            5945550   

   Other people in need of international protection  Host Community  
68                                          3582202         2304506  
69                                          3862102         4369021  
70                                          4406432         6731133  
71                                          5217456        23957770  
72                                          5755363        26095474

In [5]:
UNHCR_WC = pd.read_csv("../data/uhncr-with-countries/population.csv", skiprows=14)

In [6]:
UNHCR_WC.head()

Year       Country of origin Country of origin (ISO) Country of asylum  \
0  2015             Afghanistan                     AFG       Afghanistan   
1  2015  Iran (Islamic Rep. of)                     IRN       Afghanistan   
2  2015                Pakistan                     PAK       Afghanistan   
3  2015              Tajikistan                     TJK       Afghanistan   
4  2015                   Egypt                     EGY           Albania   

  Country of asylum (ISO)  Refugees under UNHCR's mandate  Asylum-seekers  \
0                     AFG                               0               0   
1                     AFG                              30              20   
2                     AFG                          257523              51   
3                     AFG                               0               7   
4                     ALB                               5               0   

   IDPs of concern to UNHCR Other people in need of international protection  \
0                   1174306                                                -   
1                         0                                                -   
2                         0                                                -   
3                         0                                                -   
4                         0                                                -   

   Stateless persons  Host Community  Others of concern  
0                  0               0             150317  
1                  0               0                  0  
2                  0               0                  0  
3                  0               0                  0  
4                  0               0                  0

### 2. UNRWA estimates

In [7]:
UNRWA = pd.read_csv("../data/unrwa-refugees/unrwa.csv",skiprows=14)

In [8]:
UNRWA.tail()

Year Country of origin Country of origin (ISO) Country of asylum  \
67  2019                 -                       -                 -   
68  2020                 -                       -                 -   
69  2021                 -                       -                 -   
70  2022                 -                       -                 -   
71  2023                 -                       -                 -   

   Country of asylum (ISO)    Total  
67                       -  5629829  
68                       -  5703521  
69                       -  5792907  
70                       -  5887353  
71                       -  5936247

In [9]:
UNRWA_WC = pd.read_csv("../data/unrwa-with-countries/unrwa.csv", skiprows=14)

In [10]:
UNRWA_WC.head()

Year Country of origin Country of origin (ISO)   Country of asylum  \
0  2015       Palestinian                     PSE              Jordan   
1  2015       Palestinian                     PSE             Lebanon   
2  2015       Palestinian                     PSE    Syrian Arab Rep.   
3  2015       Palestinian                     PSE  State of Palestine   
4  2016       Palestinian                     PSE              Jordan   

  Country of asylum (ISO)    Total  
0                     JOR  2144233  
1                     LBN   458369  
2                     SYR   534654  
3                     PSE  2104001  
4                     JOR  2175491

### 3. IMDC internally displaced people estimates

In [11]:
IMDC = pd.read_csv("../data/internally-displaced-idmc/idmc.csv")

In [12]:
IMDC.tail()

Year Country of origin Country of origin (ISO) Country of asylum  \
30  2019                 -                       -                 -   
31  2020                 -                       -                 -   
32  2021                 -                       -                 -   
33  2022                 -                       -                 -   
34  2023                 -                       -                 -   

   Country of asylum (ISO)     Total  
30                       -  45667305  
31                       -  48027950  
32                       -  53246765  
33                       -  62500170  
34                       -  68279000

In [13]:
IMDC_WC = pd.read_csv("../data/idmc-with-country-of-origin/idmc.csv", skiprows=14)

In [14]:
IMDC_WC.tail()

Year Country of origin Country of origin (ISO) Country of asylum  \
60  2023         Nicaragua                     NIC                 -   
61  2023      South Africa                     ZAF                 -   
62  2023     New Caledonia                     NCL                 -   
63  2023        Kazakhstan                     KAZ                 -   
64  2023   North Macedonia                     MKD                 -   

   Country of asylum (ISO)  Total  
60                       -   1000  
61                       -    620  
62                       -    150  
63                       -    120  
64                       -    110

### 4. UN population estimates

In [15]:
POP = pd.read_csv("../data/un-population-estimates.csv")

/var/folders/yl/bd_xy2c53495c2z2g0w0yrzdtddnbq/T/ipykernel_19299/4225506290.py:1: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  POP = pd.read_csv("../data/un-population-estimates.csv")


In [16]:
POP = POP[['Time', 'ISO3_code', 'Location', 'TPopulation1July']]

In [17]:
# Value from thousands to total
POP['TPopulation1July'] = POP['TPopulation1July'] * 1000

In [18]:
POP.tail()

Time ISO3_code                   Location  TPopulation1July
43467  2097       WLF  Wallis and Futuna Islands         10036.000
43468  2098       WLF  Wallis and Futuna Islands          9989.000
43469  2099       WLF  Wallis and Futuna Islands          9940.000
43470  2100       WLF  Wallis and Futuna Islands          9887.000
43471  2101       WLF  Wallis and Futuna Islands               NaN

### 5. UN's HDI and Inequality Adjusted HDI (processed by Our World in Data)

In [19]:
IHDI = pd.read_csv("../data/inequality-adjusted-human-development-index.csv")

In [20]:
IHDI.tail()

Entity Code  Year  Inequality-adjusted Human Development Index
2101  Zimbabwe  ZWE  2018                                        0.394
2102  Zimbabwe  ZWE  2019                                        0.376
2103  Zimbabwe  ZWE  2020                                        0.373
2104  Zimbabwe  ZWE  2021                                        0.370
2105  Zimbabwe  ZWE  2022                                        0.370

In [21]:
HDI = pd.read_csv("../data/human-development-index.csv")

In [22]:
HDI.head()

Entity Code  Year  Human Development Index
0  Afghanistan  AFG  1990                    0.284
1  Afghanistan  AFG  1991                    0.292
2  Afghanistan  AFG  1992                    0.299
3  Afghanistan  AFG  1993                    0.307
4  Afghanistan  AFG  1994                    0.300

### 6. DW's regional country divisions

In [23]:
DIVISIONS = pd.read_csv("../data/country-and-continent-codes-list.csv", sep=';')

In [24]:
DIVISIONS.head()

continent            region         subregion         country  M49 code  \
0      Asia     Southern Asia     Southern Asia     Afghanistan         4   
1    Europe   Southern Europe   Southern Europe         Albania         8   
2  Americas  Northern America  Northern America      Antarctica        10   
3    Africa   Northern Africa   Northern Africa         Algeria        12   
4   Oceania         Polynesia         Polynesia  American Samoa        16   

  ISO-alpha3 code ISO-alpha2 code Other groupings  
0             AFG              AF        LDC LLDC  
1             ALB              AL             NaN  
2             ATA              AQ             NaN  
3             DZA              DZ             NaN  
4             ASM              AS            SIDS

### 7. Asylum applications

In [25]:
ASA = pd.read_csv("../data/asylum-applications/asylum-applications.csv")

In [26]:
ASA.head()

Year Country of origin Country of origin (ISO) Country of asylum  \
0  2000                 -                       -                 -   
1  2000                 -                       -                 -   
2  2000                 -                       -                 -   
3  2000                 -                       -                 -   
4  2000                 -                       -                 -   

  Country of asylum (ISO) Authority Application type Stage of procedure  \
0                       -         G                V                 FI   
1                       -         U                V                 FI   
2                       -         G                V                 AR   
3                       -         G                V                 FA   
4                       -         G                V                 FI   

  Cases / Persons  applied  
0               C   109482  
1               P    71089  
2               C    54076  
3               P    33835  
4               P   564487

### 8. Asylum decisions

In [27]:
ASD = pd.read_csv("../data/asylum-decisions/asylum-decisions.csv")

In [28]:
ASD.head()

Year Country of origin Country of origin (ISO) Country of asylum  \
0  2000                 -                       -                 -   
1  2001                 -                       -                 -   
2  2002                 -                       -                 -   
3  2003                 -                       -                 -   
4  2004                 -                       -                 -   

  Country of asylum (ISO) Authority Stage of procedure Cases / Persons  \
0                       -         G                 AR               C   
1                       -         G                 AR               C   
2                       -         G                 AR               C   
3                       -         G                 AR               C   
4                       -         G                 AR               C   

   Recognized decisions  Complementary protection  Rejected decisions  \
0                  3939                         0               21349   
1                  8892                         0               40060   
2                 14070                         0               53519   
3                 15908                         0               67351   
4                 11971                         0               47039   

   Otherwise closed  Total decisions  
0              1436            26725  
1              2015            50972  
2              5425            73204  
3              4975            88234  
4              2566            61576

### 9. Refugee returns

In [29]:
RET = pd.read_csv("../data/refugee-returns/solutions.csv", skiprows=15)

In [30]:
RET.head()

Year       Country of origin Country of origin (ISO) Country of asylum  \
0  2023             Afghanistan                     AFG                 -   
1  2023                 Burundi                     BDI                 -   
2  2023    Central African Rep.                     CAF                 -   
3  2023                Cameroon                     CMR                 -   
4  2023  Dem. Rep. of the Congo                     COD                 -   

  Country of asylum (ISO)  Returned refugees  
0                       -              57530  
1                       -              26663  
2                       -              14449  
3                       -               2058  
4                       -               1808


---

## Data wrangling

### 1. Share of world population that is displaced

The count of displaced population includes:
- Refugees and people in situation akin to refugees, under the mandate of the UNHCR
- Refugees in the Palestinian Territories, under the mandate of the UNRWA
- Asylum seekers and people in akin situations, according to the UNHCR
- Internally displaced people, excluding those due to disasters, by the IDMC estimates.

The displaced population at the end of each year is then divided by the United Nations populate estimations for the 1st of July in the same year.

One important notice is that some of this numbers differ from those in the 2023 report, despite supposedly coming from the same source. Some of this can be reasonably assumed to be due to rounding up to the ten thousands slot. Others, not.

Namely, the data differs in the following entries in 2023:

- **Refugees under UNHCR's mandate**: 31,639,000 in the report x 31,637,408 in the data. (1592 less in the data)
- **Refugees under UNRWA's mandate**: 5,969,000 in the report x 5,936,247 in the data. (32573 less in the data)
- **Asylum seekers**: 6,860,000 in the report x 6,858,499 in the data.  (1501 less in the data)
- **Other people in need of protection** 5,755,000 in the report: x 5,755,363 in the data. (363 more in the data)
- **Internally displaced (IDMC)**: 68,279,000 in the report x 68,279,000 in the data. (Same values)

For the year 2023, the report also made an adjustment: it estimated that 70% of the 1.7 million of people internally displaced in Gaza were already counted under the "refugees under the URNWA's mandate" label. Thus, this share (1.7 million * 0.7) was removed from the total displaced people in the year, avoiding double counting.

When this is accounted for, the numbers we arrive at are still not the same as the report. We arrived at 117,269,517. The report says it's 117,305,000. The total difference ammounts to 28,483 people (0.02% of the total reported).


In [31]:
def share_of_world_pop(unhcr, unrwa, imdc, pop):
    
    # Create a count of the affected population in each criteria for any give year from 2013 onwards

    # Selects the valid entries from the UNHCR dataset
    unhcr_ = unhcr.copy() # Local copy to avoid in place modification
    unhcr_ = unhcr_[unhcr_.Year.isin(range(2014,2024))] # Last decade for comparisons
    unhcr_ = unhcr_[['Year', 'Refugees under UNHCR\'s mandate', 'Asylum-seekers', 
                     'Other people in need of international protection']]
    
    # Selects the refugees under the UNRWA authority (Palestinian territories)
    unrwa_ = unrwa.copy()
    unrwa_ = unrwa_[unrwa_.Year.isin(range(2014,2024))]
    unrwa_ = unrwa_[['Year', 'Total']]
    unrwa_ = unrwa_.rename(columns={'Total': 'Refugees under UNRWA\'s mandate'})
    
    # Selects the internally displaced people according to the IMDC data
    imdc_ = imdc.copy()
    imdc_ = imdc_[imdc_.Year.isin(range(2014,2024))]
    imdc_ = imdc_[['Year', 'Total']]
    imdc_ = imdc_.rename(columns={'Total': 'IDP\'s according to IMDC'})
    
    # Selects the population estimates for each year
    pop_ = pop.copy()
    pop_ = pop_[pop_.Time.isin(range(2014,2024))]
    pop_ = pop_[pop_.Location=='World']
    pop_ = pop_[['Time', 'TPopulation1July']]
    pop_ = pop_.rename(columns={"Time": "Year", "TPopulation1July":"Population"})
    
    
    # Merges everything by year
    data = unhcr_.merge(unrwa_).merge(imdc_).merge(pop_)
    data = data.replace("-", 0).astype(float)
        
    # Computes total population displaced
    data['Displaced population'] = data['Refugees under UNHCR\'s mandate'] + data['Asylum-seekers'] + \
                                   data['Other people in need of international protection'] + \
                                   data['Refugees under UNRWA\'s mandate'] + data['IDP\'s according to IMDC']
    
    # In 2023, we have to make an adjustment to account for a methodology change than was described
    # here: https://www.unhcr.org/refugee-statistics/insights/explainers/forcibly-displaced-pocs.html
    # "At the end of 2023, UNRWA estimates that 70 per cent of the 1.7 million IDPs in the Gaza Strip 
    # at end-2023 were Palestine refugees under its mandate. These internally displaced refugees under 
    # UNRWA’s mandate are only counted once in the global forcibly displaced total."
    # So, for 2023, we will remove 1.7 million * 0.7 from the total displaced population
    data.loc[data.Year==2023, 'Displaced population'] = data.loc[data.Year==2023, 'Displaced population'] - (1700000 * 0.7)
        
    # Compute the share of people per thousand that are displaced
    data['Displaced per thousand'] = (data['Displaced population'] / data['Population']) * 1000
    
    # Formats year
    data['Year'] = data.Year.astype(int)
    
    # Display, save and return variable
    display(data)
    data.to_csv("../output/Line chart – share of population forcibly displaced.csv", index=False)
    return data
    
    

In [32]:
DATA = share_of_world_pop(UNHCR, UNRWA, IMDC, POP)

Year  Refugees under UNHCR's mandate  Asylum-seekers  \
0  2014                    14384289.000     1794704.000   
1  2015                    16110276.000     3223460.000   
2  2016                    17184286.000     2729521.000   
3  2017                    19940566.000     3089507.000   
4  2018                    20359553.000     3501629.000   
5  2019                    20414669.000     4148141.000   
6  2020                    20661846.000     4184926.000   
7  2021                    21327285.000     4616135.000   
8  2022                    29429078.000     5442319.000   
9  2023                    31637408.000     6858499.000   

   Other people in need of international protection  \
0                                             0.000   
1                                             0.000   
2                                             0.000   
3                                             0.000   
4                                       2592947.000   
5                                       3582202.000   
6                                       3862102.000   
7                                       4406432.000   
8                                       5217456.000   
9                                       5755363.000   

   Refugees under UNRWA's mandate  IDP's according to IMDC     Population  \
0                     5149742.000             37877320.000 7339013419.000   
1                     5241257.000             40451900.000 7426597537.000   
2                     5340443.000             40220850.000 7513474238.000   
3                     5442947.000             39934042.000 7599822404.000   
4                     5545538.000             41312940.000 7683789828.000   
5                     5629829.000             45667305.000 7764951032.000   
6                     5703521.000             48027950.000 7840952880.000   
7                     5792907.000             53246765.000 7909295152.000   
8                     5887353.000             62500170.000 7975105156.000   
9                     5936247.000             68279000.000 8045311448.000   

   Displaced population  Displaced per thousand  
0          59206055.000                   8.067  
1          65026893.000                   8.756  
2          65475100.000                   8.714  
3          68407062.000                   9.001  
4          73312607.000                   9.541  
5          79442146.000                  10.231  
6          82440345.000                  10.514  
7          89389524.000                  11.302  
8         108476376.000                  13.602  
9         117276517.000                  14.577

In [33]:
DATA.Year

0    2014
1    2015
2    2016
3    2017
4    2018
5    2019
6    2020
7    2021
8    2022
9    2023
Name: Year, dtype: int64

### 2. Of all the displaced people, how many are internally displaced?

Here, we will use the 2023 data only. 

We will also need to account for the "double conting" of Palestinian refugees. To do so, we will remove 1.19 millon million (1.7 million * 0.7) from the refugee total, as explained above, so they are not doubly conted.

In [34]:
def donut_chart(data):
    
    # Total displaced
    total_displaced = int(data.loc[data.Year==2023]['Displaced population'])

    # Total refugees
    unhcr_refugees = int(data.loc[data.Year==2023]['Refugees under UNHCR\'s mandate'])
    unrwa_refugees = int(data.loc[data.Year==2023]['Refugees under UNRWA\'s mandate'] - (1700000 * 0.7))
    other_people = int(data.loc[data.Year==2023]['Other people in need of international protection'])
    asylum_seekers = int(data.loc[data.Year==2023]['Asylum-seekers'])

    # Total internally displaced population
    internally_displaced = int(data.loc[data.Year==2023]['IDP\'s according to IMDC'])  
    
    # Total refugees
    # UNHCR's mandate + Other people in need of international protection + UNRWA's mandate - 1.19 million double counted
    total_refugees = unhcr_refugees + unrwa_refugees + other_people
    
    # Check if the math we did is alright – that is, all displaced people are included in the sum
    assert(total_refugees + internally_displaced + asylum_seekers == total_displaced)
    
    # Dictionary to store the data
    shares = {}
    shares['Internally displaced'] = round(internally_displaced / total_displaced * 100, 2)
    shares['Refugees and people in need of international protection'] = round(total_refugees / total_displaced * 100, 2)
    shares['Asylum seekers'] = round(asylum_seekers / total_displaced * 100, 2)
    
    assert (internally_displaced + total_refugees + asylum_seekers) == (total_displaced)
    
    to_export = pd.DataFrame.from_dict(shares, orient='index').reset_index()
    to_export.to_csv("../output/pie-chart-shares-each-type.csv", index=False)

In [35]:
donut_chart(DATA)

### 3. Where are the internally displaced people?

In [36]:
def internally_displaced_map(imdc_wc, pop):
    
    # Make copies to avoid changing original variables
    imdc_wc_ = imdc_wc.copy()
    pop_ = pop.copy()
    
    # Keeps only 2023 population
    pop_ = pop_[pop_.Time==2023]
    
    # Rename columns for compatibility
    pop_ = pop_.rename(columns={
        "Time": "Year",
        "ISO3_code": "Country of origin (ISO)",
        "TPopulation1July": "Population"
    })
    
    # Removes entries that are not countries (no ISO3 code)
    pop_[~pop_['Country of origin (ISO)'].isna()]
    
    # Merge
    merged = imdc_wc_.merge(pop_)
    
    # There's a metodological decision to make: around 42,000 people are internally displaced
    # in the so called Abyei Area, which is contested between Sudan and South Sudan.
    # Instead of allocating those to one or another, we will exclude them from the data visualization.
    # The number of displaced people there is roughly 0.46% of the IDPs in Sudan, 3.7% of the IDPs
    # in South Sudan and 0.06% of the displaced people worlwide.

    # Makes sure we have all the entries with IDPs
    assert merged.shape[0] == imdc_wc_.shape[0] - 1 # Asserts only one entry is missing (Abyei Area)
    
    # More column renaming for clarity
    merged = merged.rename(columns={"Total": "IDPs"})
    
    # Remove not needed columns
    merged = merged[['Location', 'Country of origin (ISO)', 'IDPs', 'Population']]
    
    # Computes a ratio
    merged['Internally displaced per 100 people'] = merged['IDPs'] / merged['Population'] * 100
    
    # Sorts
    merged = merged.sort_values(by='Internally displaced per 100 people', ascending=False)

    # Save to CSV
    merged.to_csv("../output/Symbol map – internally displaced people.csv", index=False)
    
    return merged

In [37]:
DATA_IDP = internally_displaced_map(IMDC_WC, POP)

### 4. Table with share of forcibly displaced people in the population

We will also create a datwarapper table that shows the share of total population in each country that is forcibly displaced.

In [38]:
def idp_table(data_idp):
    
    # Copies locally
    data_idp_ = data_idp.copy()
    
    # Top 20 countries by share of people displaced
    data_idp_ = data_idp_.sort_values(by='Internally displaced per 100 people', ascending=False).head(20)
    
    # Get ISO2
    data_idp_['ISO2'] = data_idp_['Country of origin (ISO)'].apply(lambda x: pycountry.countries.get(alpha_3=x).alpha_2)
    
    # Create a display name
    data_idp_['Display name'] = ":" + data_idp_.ISO2 + ": " + data_idp_['Location']
    data_idp_['Display name'] = data_idp_['Display name'].str.replace("State of Palestine", "Palestinian Territories")
    data_idp_['Display name'] = data_idp_['Display name'].str.replace("Syrian Arab Republic", "Syria")
    data_idp_['Display name'] = data_idp_['Display name'].str.replace("Democratic Republic of the Congo", "DR Congo")

    # Reorder and filters columns
    data_idp_ = data_idp_[['Display name', 'Internally displaced per 100 people']]
    
    data_idp_.to_csv("../output/Table – IDP per 100 inhabitants.csv", index=False)
    
    return data_idp_

In [39]:
idp_table(DATA_IDP)

Display name  Internally displaced per 100 people
12   :PS: Palestinian Territories                               31.836
1                      :SY: Syria                               31.205
6                    :SO: Somalia                               21.286
28                    :CY: Cyprus                               19.680
0                      :SD: Sudan                               18.818
4                      :YE: Yemen                               13.109
14               :SS: South Sudan                               10.109
7                    :UA: Ukraine                               10.040
5                :AF: Afghanistan                                9.912
3                   :CO: Colombia                                9.747
20  :CF: Central African Republic                                8.916
11              :BF: Burkina Faso                                8.873
27                   :GE: Georgia                                8.342
2                   :CD: DR Congo                                6.585
17                :AZ: Azerbaijan                                6.319
10                   :MM: Myanmar                                4.810
16                  :CM: Cameroon                                3.644
30                    :RS: Serbia                                2.937
35    :BA: Bosnia and Herzegovina                                2.834
26                     :HT: Haiti                                2.653

In [40]:
DIVISIONS

continent              region         subregion  \
0        Asia       Southern Asia     Southern Asia   
1      Europe     Southern Europe   Southern Europe   
2    Americas    Northern America  Northern America   
3      Africa     Northern Africa   Northern Africa   
4     Oceania           Polynesia         Polynesia   
..        ...                 ...               ...   
244  Americas       South America     South America   
245   Oceania           Polynesia         Polynesia   
246   Oceania           Polynesia         Polynesia   
247      Asia        Western Asia      Western Asia   
248    Africa  Sub-Saharan Africa    Eastern Africa   

                                country  M49 code ISO-alpha3 code  \
0                           Afghanistan         4             AFG   
1                               Albania         8             ALB   
2                            Antarctica        10             ATA   
3                               Algeria        12             DZA   
4                        American Samoa        16             ASM   
..                                  ...       ...             ...   
244  Venezuela (Bolivarian Republic of)       862             VEN   
245           Wallis and Futuna Islands       876             WLF   
246                               Samoa       882             WSM   
247                               Yemen       887             YEM   
248                              Zambia       894             ZMB   

    ISO-alpha2 code Other groupings  
0                AF        LDC LLDC  
1                AL             NaN  
2                AQ             NaN  
3                DZ             NaN  
4                AS            SIDS  
..              ...             ...  
244              VE             NaN  
245              WF             NaN  
246              WS            SIDS  
247              YE             LDC  
248              ZM        LDC LLDC  

[249 rows x 8 columns]

In [41]:
### Share of IDPs by continent
by_continent = IMDC_WC.merge(DIVISIONS, left_on='Country of origin (ISO)', right_on='ISO-alpha3 code')

In [42]:
#### Continents
by_continent.groupby('continent').Total.sum() / by_continent.groupby('continent').Total.sum().sum()

continent
Africa     0.477
Americas   0.092
Asia       0.371
Europe     0.059
Oceania    0.001
Name: Total, dtype: float64

In [43]:
#### Middle East countries
by_continent[ (by_continent.subregion=='Western Asia') &
             (~by_continent['Country of origin (ISO)'].isin(['AZE', 'GEO', 'CYP', 'ARM', 'TUR']))]

Year Country of origin Country of origin (ISO) Country of asylum  \
1   2023  Syrian Arab Rep.                     SYR                 -   
4   2023             Yemen                     YEM                 -   
12  2023       Palestinian                     PSE                 -   
13  2023              Iraq                     IRQ                 -   
31  2023            Israel                     ISR                 -   
38  2023           Lebanon                     LBN                 -   

   Country of asylum (ISO)    Total continent        region     subregion  \
1                        -  7248000      Asia  Western Asia  Western Asia   
4                        -  4516000      Asia  Western Asia  Western Asia   
12                       -  1710000      Asia  Western Asia  Western Asia   
13                       -  1124000      Asia  Western Asia  Western Asia   
31                       -   200000      Asia  Western Asia  Western Asia   
38                       -    74000      Asia  Western Asia  Western Asia   

                 country  M49 code ISO-alpha3 code ISO-alpha2 code  \
1   Syrian Arab Republic       760             SYR              SY   
4                  Yemen       887             YEM              YE   
12    State of Palestine       275             PSE              PS   
13                  Iraq       368             IRQ              IQ   
31                Israel       376             ISR              IL   
38               Lebanon       422             LBN              LB   

   Other groupings  
1              NaN  
4              LDC  
12             NaN  
13             NaN  
31             NaN  
38             NaN

In [44]:
#### Share of Middle East in total
by_continent[ (by_continent.subregion=='Western Asia') &
             (~by_continent['Country of origin (ISO)'].isin(['AZE', 'GEO', 'CYP', 'ARM', 'TUR']))].Total.sum() / \
by_continent.Total.sum()

0.217946275481044

### 5. Map with the countries of origin of refuguees

To make the map, we will need, again, to combine the figures from UNHCR and UNRWA. We are, this time, keeping the UNRWA's refugees that were relocated _within_ Palestine, as in the report they are considered to have refugee status.

In [45]:
def refugee_map_source(unhcr_wc, unrwa_wc):
    
    # Local copies
    unhcr_wc_ = unhcr_wc.copy()
    unrwa_wc_ = unrwa_wc.copy()
    
    # Keep only 2023 data
    unhcr_wc_ = unhcr_wc_[unhcr_wc_.Year==2023]
    unrwa_wc_ = unrwa_wc_[unrwa_wc_.Year==2023]
    
    # Changes the - for 0 when applicable
    unhcr_wc_['Other people in need of international protection'] = unhcr_wc_['Other people in need of international protection']\
                                                                              .str.replace('-', '0').astype(int)
    
    # Keep only the refugees on the UNHCR dataset
    unhcr_wc_ = unhcr_wc_[['Country of origin', 'Country of origin (ISO)', 
                          'Refugees under UNHCR\'s mandate', 'Other people in need of international protection']]
    
    # Combines them by country of origin
    unhcr_wc_ = unhcr_wc_.groupby(['Country of origin', 'Country of origin (ISO)']).sum().reset_index()
    
    # Does the same for unrwa_wc_
    unrwa_wc_ = unrwa_wc_[['Country of origin', 'Country of origin (ISO)', 'Total']]
    unrwa_wc_ = unrwa_wc_.rename(columns={"Total": "Refugees under UNRWA's mandate"})
    unrwa_wc_ = unrwa_wc_.groupby(['Country of origin', 'Country of origin (ISO)']).sum().reset_index()
    
    
    # Joins them both
    joined = pd.concat([unrwa_wc_, unhcr_wc_])
    
    # Regroups again to avoid two Palestinian entries
    joined = joined.groupby(['Country of origin', 'Country of origin (ISO)']).sum().reset_index()
    
    # Creates a sum of all refugees
    joined['Total refugees'] = joined.iloc[: , 2:].sum(axis=1) # Sums all the columns from index 2 onwards
    
    # Keeps only that
    joined = joined[['Country of origin', 'Country of origin (ISO)', 'Total refugees']]
    
    # Sorts
    joined = joined.sort_values(by='Total refugees', ascending=False)
    
    # Fixes display names
    joined['Country of origin'] = joined['Country of origin'].str.replace("Syrian Arab Rep.", "Syria")
    joined['Country of origin'] = joined['Country of origin'].str.replace("Venezuela \(Bolivarian Republic of\)", "Venezuela")
    joined['Country of origin'] = joined['Country of origin'].str.replace("Palestinian", "Palestinian Territories")
    joined['Country of origin'] = joined['Country of origin'].str.replace("Dem. Rep. of the Congo", "DR Congo")
    joined['Country of origin'] = joined['Country of origin'].str.replace("Iran \(Islamic Rep. of\)", "Iran")
    joined['Country of origin'] = joined['Country of origin'].str.replace("Bolivia \(Plurinational State of\)", "Bolivia")


    # Keeps only those with at least one refugee
    joined = joined[joined['Total refugees'] > 0]
    
    joined.to_csv("../output/Symbol map – Refugees originating from each country.csv", index=False)
    
    return joined

In [46]:
DATA_REFUGEE_SRC = refugee_map_source(UNHCR_WC, UNRWA_WC)

In [47]:
DATA_REFUGEE_SRC.head(10)

Country of origin Country of origin (ISO)  Total refugees
0                Afghanistan                     AFG     6403144.000
176                    Syria                     SYR     6355788.000
199                Venezuela                     VEN     6103058.000
137  Palestinian Territories                     PSE     5977324.000
190                  Ukraine                     UKR     5960362.000
168              South Sudan                     SSD     2292482.000
172                    Sudan                     SDN     1496923.000
122                  Myanmar                     MMR     1283426.000
51                  DR Congo                     COD      978209.000
166                  Somalia                     SOM      842044.000

In [48]:
DATA_REFUGEE_SRC.head(10)['Total refugees'].sum() / DATA_REFUGEE_SRC['Total refugees'].sum()

0.869919553681092

### 6. Map with the countries that take in the most refugees

To make this map, we will follow the same approach as the previous one, but instead of grouping by Country of Origin, we will group by Country of Asylum.

One methodological question remains: there are people officially considered as refugees in the Palestinian Territories, originating from there as well. This is because of the exceptionalities of the UNRWA's refugees.

In order to visualize the data comaprably, focusing in countries which are mostly absorbing refugees from elsewhere, we will remove those from the map.

In [49]:
def refugee_map_dest(unhcr_wc, unrwa_wc):
    
    # Local copies
    unhcr_wc_ = unhcr_wc.copy()
    unrwa_wc_ = unrwa_wc.copy()
    
    # Keep only 2023 data
    unhcr_wc_ = unhcr_wc_[unhcr_wc_.Year==2023]
    unrwa_wc_ = unrwa_wc_[unrwa_wc_.Year==2023]
    
    # Changes the - for 0 when applicable
    unhcr_wc_['Other people in need of international protection'] = unhcr_wc_['Other people in need of international protection']\
                                                                              .str.replace('-', '0').astype(int)
    
    # Keep only the refugees on the UNHCR dataset
    unhcr_wc_ = unhcr_wc_[['Country of asylum', 'Country of asylum (ISO)', 
                          'Refugees under UNHCR\'s mandate', 'Other people in need of international protection']]
    
    # Combines them by country of origin
    unhcr_wc_ = unhcr_wc_.groupby(['Country of asylum', 'Country of asylum (ISO)']).sum().reset_index()
    
    # Does the same for unrwa_wc_
    unrwa_wc_ = unrwa_wc_[['Country of asylum', 'Country of asylum (ISO)', 'Total']]
    unrwa_wc_ = unrwa_wc_.rename(columns={"Total": "Refugees under UNRWA's mandate"})
    unrwa_wc_ = unrwa_wc_.groupby(['Country of asylum', 'Country of asylum (ISO)']).sum().reset_index()
    
    # Removes the Palestinian refugees originating from the Palestinian territories
    unrwa_wc_ = unrwa_wc_[unrwa_wc_['Country of asylum (ISO)']!='PSE']
    
    # Joins them both
    joined = pd.concat([unrwa_wc_, unhcr_wc_])
    
    # Regroups again to avoid two Palestinian entries
    joined = joined.groupby(['Country of asylum', 'Country of asylum (ISO)']).sum().reset_index()
    
    # Creates a sum of all refugees
    joined['Total refugees'] = joined.iloc[: , 2:].sum(axis=1) # Sums all the columns from index 2 onwards
    
    # Keeps only that
    joined = joined[['Country of asylum', 'Country of asylum (ISO)', 'Total refugees']]
    
    # Sorts
    joined = joined.sort_values(by='Total refugees', ascending=False)
    
    # Fixes display names
    joined['Country of asylum'] = joined['Country of asylum'].str.replace("Syrian Arab Rep.", "Syria")
    joined['Country of asylum'] = joined['Country of asylum'].str.replace("Venezuela \(Bolivarian Republic of\)", "Venezuela")
    joined['Country of asylum'] = joined['Country of asylum'].str.replace("Palestinian", "Palestinian Territories")
    joined['Country of asylum'] = joined['Country of asylum'].str.replace("Dem. Rep. of the Congo", "DR Congo")
    joined['Country of asylum'] = joined['Country of asylum'].str.replace("Iran \(Islamic Rep. of\)", "Iran")
    joined['Country of asylum'] = joined['Country of asylum'].str.replace("Bolivia \(Plurinational State of\)", "Bolivia")

    
    # For each one of the countries, we will also store information about the country of origin
    # of the top 5 refugee populations they host
    for index, row in joined.iterrows():
        
        # Selects the country
        country = row['Country of asylum']
        iso3 = row['Country of asylum (ISO)']
        
#         print(country, iso3)
        
        # Access the information in the original dataset – UNHCR
        subset_unhcr = unhcr_wc[(unhcr_wc.Year==2023) & (unhcr_wc['Country of asylum (ISO)']==iso3)]
        
        # Access the information in the original datset – UNRWA
        subset_unrwa = unrwa_wc[(unrwa_wc.Year==2023) & (unrwa_wc['Country of asylum (ISO)']==iso3)]
        
        # Computes the total UNHCR refugees
        subset_unhcr['Other people in need of international protection'] = subset_unhcr['Other people in need of international protection'].str.replace('-', '0').astype(int)
        subset_unhcr = subset_unhcr[['Country of origin', 'Country of origin (ISO)', 
                              'Refugees under UNHCR\'s mandate', 'Other people in need of international protection']]
        
        # Renames columns in the unrwa subset
        subset_unrwa = subset_unrwa[['Country of origin', 'Country of origin (ISO)', 'Total']]
        subset_unrwa = subset_unrwa.rename(columns={"Total": "Refugees under UNRWA's mandate"})
        
        # Joins them both
        subset_joined = pd.concat([subset_unhcr, subset_unrwa])
        
        # Groups by to avoid two Palestinian entries (because of the UNRWA)
        subset_joined = subset_joined.groupby(['Country of origin', 'Country of origin (ISO)']).sum().reset_index()
        
        # Joined total
        subset_joined['Total refugees'] = subset_joined.iloc[: , 2:].sum(axis=1)
        
        # Gets the country with the highest refugee count and creates a string out of it
        subset_joined = subset_joined.sort_values(by='Total refugees', ascending=False)
        
#         if iso3 == 'JOR':
#             display(subset_joined)
#             return     
        
#         if iso3 == 'DEU':
#             display(subset_joined.head(20))
        
        # Adds back to the original data
        joined.loc[index, 'Top origin'] = subset_joined.iloc[0]['Country of origin']
        joined.loc[index, 'Refugees from top origin'] = subset_joined.iloc[0]['Total refugees']
    
    # Keeps only places that received refugees
    joined = joined[joined['Total refugees'] > 0]
    
    # Groups by to avoid having two separate entries for palestine
    joined.to_csv("../output/Symbol map – Refugees relocating to each country.csv", index=False)
    
    return joined

In [50]:
DATA_REFUGEE_DEST = refugee_map_dest(UNHCR_WC, UNRWA_WC)

In [51]:
DATA_REFUGEE_DEST.head(10)

Country of asylum Country of asylum (ISO)  Total refugees  \
76                 Iran                     IRN     3764517.000   
162             Türkiye                     TUR     3251127.000   
82               Jordan                     JOR     3063591.000   
36             Colombia                     COL     2852542.000   
63              Germany                     DEU     2593007.000   
120            Pakistan                     PAK     1988231.000   
163              Uganda                     UGA     1577498.000   
88              Lebanon                     LBN     1279108.000   
132  Russian Federation                     RUS     1230131.000   
32                 Chad                     TCD     1100921.000   

                             Top origin  Refugees from top origin  
76                          Afghanistan               3752317.000  
162                    Syrian Arab Rep.               3214780.000  
82                          Palestinian               2379525.000  
36   Venezuela (Bolivarian Republic of)               2852294.000  
63                              Ukraine               1097503.000  
120                         Afghanistan               1987717.000  
163                         South Sudan                923607.000  
88                     Syrian Arab Rep.                784884.000  
132                             Ukraine               1227554.000  
32                                Sudan                923323.000

### 7. Developing countries take the most refugees

To show how least developed countries shoulder a disporpotionate share of refugees, I will make a swarm plot
plotting the % of refugees in the country population x their HDI.

In [52]:
DATA_REFUGEE_DEST[DATA_REFUGEE_DEST['Country of asylum (ISO)']=='JOR']

Country of asylum Country of asylum (ISO)  Total refugees   Top origin  \
82            Jordan                     JOR     3063591.000  Palestinian   

    Refugees from top origin  
82               2379525.000

In [53]:
def scatter_plot(data_refugee_dest, pop, hdi, divisions):
    
    # Make local copies
    data_refugee_dest_ = data_refugee_dest.copy()
    pop_ = pop.copy()
    hdi_ = hdi.copy()
    divisions_ = divisions.copy()
    
    # Keep only most recent years
    pop_ = pop_[pop_['Time']==2023]
    hdi_ = hdi_[hdi_['Year']==2022]
    
    # Clean columns we don't need
    data_refugee_dest_ = data_refugee_dest_.drop(columns=['Top origin', 'Refugees from top origin'])
    pop_ = pop_[['TPopulation1July', 'ISO3_code']]
    hdi_ = hdi_[['Code', 'Human Development Index']]
    divisions_ = divisions_[['continent', 'ISO-alpha3 code']]
    

    # Standardizes columns names
    data_refugee_dest_ = data_refugee_dest_.rename(columns={'Country of asylum (ISO)': 'ISO3', 'Country of asylum': 'Country'})
    pop_ = pop_.rename(columns={'TPopulation1July':'Population', 'ISO3_code': 'ISO3'})
    hdi_ = hdi_.rename(columns={'Code':'ISO3', 'Human Development Index': 'HDI'})
    divisions_ = divisions_.rename(columns={"ISO-alpha3 code": "ISO3", "continent":"Continent"})
    
    # Join by ISO3 code
    merged = data_refugee_dest_.merge(pop_).merge(hdi_).merge(divisions_)
    
    # Compute share of refugees in the population
    merged['Refugees in population, per 1000'] = merged['Total refugees'] / merged['Population'] * 1000
    
    # Add buckets for HDI levels
    # Using this as reference: https://hdr.undp.org/data-center/human-development-index#/indicies/HDI
    def bin_hdi(value):
        if value > 0.8:
            return 'very high'
        elif value > 0.7:
            return 'high'
        elif value > 0.55:
            return 'medium'
        else:
            return 'low'
        
    merged['HDI Level'] = merged['HDI'].apply(bin_hdi)
    
    merged.to_csv("../output/Scatter – refugees x HDI.csv", index=False)
    return merged


In [54]:
REFUGEES_HDI = scatter_plot(DATA_REFUGEE_DEST, POP, HDI, DIVISIONS)

In [55]:
REFUGEES_HDI.sort_values(by='Refugees in population, per 1000', ascending=False).head(20)

Country ISO3  Total refugees   Population   HDI Continent  \
2            Jordan  JOR     3063591.000 11337052.000 0.736      Asia   
7           Lebanon  LBN     1279108.000  5353930.000 0.723      Asia   
64       Montenegro  MNE       64733.000   626484.000 0.844    Europe   
9              Chad  TCD     1100921.000 18278568.000 0.394    Africa   
3          Colombia  COL     2852542.000 52085168.000 0.758  Americas   
43          Armenia  ARM      150080.000  2777970.000 0.786      Asia   
0              Iran  IRN     3764517.000 89172767.000 0.780      Asia   
1           Türkiye  TUR     3251127.000 85816199.000 0.855      Asia   
27          Czechia  CZE      377120.000 10495295.000 0.895    Europe   
47  Rep. of Moldova  MDA      120947.000  3435931.000 0.763    Europe   
26      South Sudan  SSD      381236.000 11088796.000 0.381    Africa   
6            Uganda  UGA     1577498.000 48582334.000 0.550    Africa   
72          Estonia  EST       42005.000  1322765.000 0.899    Europe   
4           Germany  DEU     2593007.000 83294633.000 0.950    Europe   
75           Cyprus  CYP       39044.000  1260138.000 0.907      Asia   
10             Peru  PER     1050334.000 34352719.000 0.762  Americas   
58        Lithuania  LTU       80225.000  2718352.000 0.879    Europe   
19          Ecuador  ECU      532340.000 18190484.000 0.765  Americas   
31          Austria  AUT      257811.000  8958960.000 0.926    Europe   
91           Guyana  GUY       21695.000   813834.000 0.742  Americas   

    Refugees in population, per 1000  HDI Level  
2                            270.228       high  
7                            238.910       high  
64                           103.327  very high  
9                             60.230        low  
3                             54.767       high  
43                            54.025       high  
0                             42.216       high  
1                             37.885  very high  
27                            35.932  very high  
47                            35.201       high  
26                            34.380        low  
6                             32.471        low  
72                            31.755  very high  
4                             31.131  very high  
75                            30.984  very high  
10                            30.575       high  
58                            29.512  very high  
19                            29.265       high  
31                            28.777  very high  
91                            26.658       high

In [56]:
# Share of total refugees by group
REFUGEES_HDI.groupby('HDI Level')['Total refugees'].sum() / REFUGEES_HDI['Total refugees'].sum()

HDI Level
high        0.352
low         0.213
medium      0.083
very high   0.352
Name: Total refugees, dtype: float64

In [57]:
REFUGEES_HDI

Country ISO3  Total refugees   Population   HDI Continent  \
0         Iran  IRN     3764517.000 89172767.000 0.780      Asia   
1      Türkiye  TUR     3251127.000 85816199.000 0.855      Asia   
2       Jordan  JOR     3063591.000 11337052.000 0.736      Asia   
3     Colombia  COL     2852542.000 52085168.000 0.758  Americas   
4      Germany  DEU     2593007.000 83294633.000 0.950    Europe   
..         ...  ...             ...          ...   ...       ...   
155       Fiji  FJI          15.000   936376.000 0.729   Oceania   
156   Mongolia  MNG          10.000  3447156.000 0.741      Asia   
157    Comoros  COM          10.000   852075.000 0.586    Africa   
158   Barbados  BRB           5.000   281996.000 0.809  Americas   
159  Mauritius  MUS           5.000  1300557.000 0.796    Africa   

     Refugees in population, per 1000  HDI Level  
0                              42.216       high  
1                              37.885  very high  
2                             270.228       high  
3                              54.767       high  
4                              31.131  very high  
..                                ...        ...  
155                             0.016       high  
156                             0.003       high  
157                             0.012     medium  
158                             0.018  very high  
159                             0.004       high  

[160 rows x 8 columns]

In [58]:
# Which are missing?
mask = ~DATA_REFUGEE_DEST['Country of asylum (ISO)'].isin(REFUGEES_HDI.ISO3)
DATA_REFUGEE_DEST[mask]

Country of asylum Country of asylum (ISO)  Total refugees  \
6                        Aruba                     ABW       17085.000   
43                    Curacao                      CUW       14000.000   
30              Cayman Islands                     CYM          49.000   
103                     Monaco                     MCO          17.000   
161   Turks and Caicos Islands                     TCA           7.000   
141  Sint Maarten (Dutch part)                     SXM           5.000   

                             Top origin  Refugees from top origin  
6    Venezuela (Bolivarian Republic of)                 17085.000  
43   Venezuela (Bolivarian Republic of)                 14000.000  
30                                 Cuba                    49.000  
103                            Viet Nam                     7.000  
161                           Sri Lanka                     7.000  
141                            Colombia                     5.000

### 8. Gap between asylum decisions and applications

In order to make this chart, we will simply take the totals reported in each year from 2014 onwards in each dataset.

In [59]:
def asylum_gap(asa, asd):
    
    # Local copies
    asa_ = asa.copy()
    asd_ = asd.copy()
    
    # Keeps only last 10 years
    asa_ = asa_[asa_.Year>=2014]
    asd_ = asd_[asd_.Year>=2014]
    
    # Keep only the count of persons
    asa_ = asa_[asa_['Cases / Persons']=='P']
    asd_ = asd_[asd_['Cases / Persons']=='P']
    
    # Keep only substantive decisions (refugee status granted or rejected), removing 
    # the number applications closed for administrative reasons, as instructed here:
    # https://www.unhcr.org/refugee-statistics/insights/explainers/forcibly-displaced-api.html
    asd_['Substantive decisions'] = asd_['Total decisions'] - asd_['Otherwise closed']
    
    # Keep only new applications
    asa_ = asa_[asa_['Application type'].isin(['N', 'NR', 'NA'])]
    
    # Group by year and sum
    asa_ = asa_.groupby('Year')['applied'].sum().reset_index()
    asd_ = asd_.groupby('Year')['Substantive decisions', 'Rejected decisions', 'Recognized decisions', 'Complementary protection', 'Otherwise closed', 'Total decisions'].sum().reset_index()
        
    # Joins
    merged = asa_.merge(asd_)
    
    # Renames columns
    merged = merged.rename(columns={"applied": "New applications"})
    
    merged.to_csv("../output/Line chart – Asylum applications x substantive decisions.csv", index=False)
    
    return merged

In [60]:
asylum_gap(ASA, ASD)

Year  New applications  Substantive decisions  Rejected decisions  \
0  2014           1379932                1010226              412064   
1  2015           2024313                1149626              488800   
2  2016           1826978                1477485              594081   
3  2017           1499626                1428885              720625   
4  2018           1584142                1067766              592827   
5  2019           1870476                1144619              603133   
6  2020            984843                 967160              517810   
7  2021           1401457                1020671              526889   
8  2022           2574411                1288530              573680   
9  2023           3580371                1404143              580987   

   Recognized decisions  Complementary protection  Otherwise closed  \
0                261394                    336768            333527   
1                423371                    237455            440314   
2                548403                    335001           1318790   
3                460648                    247612            494676   
4                326089                    148849            483181   
5                417103                    123103            476849   
6                341096                    108794            362361   
7                379376                    114406            397509   
8                533725                    181125            751358   
9                612246                    210910           1092543   

   Total decisions  
0          1343753  
1          1589940  
2          2796275  
3          1923561  
4          1550947  
5          1621468  
6          1329521  
7          1418180  
8          2039888  
9          2496686

### 9. Refugee returns

Simply map out the total refugees returning to each country.

In [61]:
def map_returns(ret):
    
    # Local copy
    ret_ = ret.copy()
    
    # Keep only the columns we need
    ret_ = ret_[['Year', 'Country of origin',  'Country of origin (ISO)', 'Returned refugees']]
    
    # Saves
    ret_.to_csv("../output/Map – Returned refugees.csv", index=False)
    
    return ret_

In [62]:
RET_REF = map_returns(RET)

In [63]:
RET_REF.head()

Year       Country of origin Country of origin (ISO)  Returned refugees
0  2023             Afghanistan                     AFG              57530
1  2023                 Burundi                     BDI              26663
2  2023    Central African Rep.                     CAF              14449
3  2023                Cameroon                     CMR               2058
4  2023  Dem. Rep. of the Congo                     COD               1808

In [64]:
RET_REF.sort_values(by='Returned refugees', ascending=False)

Year                     Country of origin Country of origin (ISO)  \
21  2023                           South Sudan                     SSD   
25  2023                               Ukraine                     UKR   
0   2023                           Afghanistan                     AFG   
24  2023                      Syrian Arab Rep.                     SYR   
15  2023                               Nigeria                     NGA   
1   2023                               Burundi                     BDI   
23  2023                                 Sudan                     SDN   
2   2023                  Central African Rep.                     CAF   
6   2023                              Ethiopia                     ETH   
3   2023                              Cameroon                     CMR   
18  2023                                Rwanda                     RWA   
19  2023                               Somalia                     SOM   
4   2023                Dem. Rep. of the Congo                     COD   
13  2023                                  Mali                     MLI   
12  2023                             Sri Lanka                     LKA   
10  2023                                  Iraq                     IRQ   
16  2023                              Pakistan                     PAK   
7   2023                         Cote d'Ivoire                     CIV   
9   2023                Iran (Islamic Rep. of)                     IRN   
20  2023  Serbia and Kosovo: S/RES/1244 (1999)                     SRB   
27  2023                                 Yemen                     YEM   
14  2023                               Myanmar                     MMR   
22  2023                             Stateless                     XXA   
26  2023                        Western Sahara                     ESH   
17  2023                    Russian Federation                     RUS   
11  2023                                 Libya                     LBY   
8   2023                             Indonesia                     IDN   
5   2023                              Colombia                     COL   

    Returned refugees  
21             527206  
25             324573  
0               57530  
24              37552  
15              31747  
1               26663  
23              17238  
2               14449  
6                6187  
3                2058  
18               1961  
19               1876  
4                1808  
13                574  
12                346  
10                154  
16                 36  
7                  27  
9                  24  
20                 12  
27                 11  
14                 10  
22                  6  
26                  6  
17                  5  
11                  5  
8                   5  
5                   5